File: codeforces/sakurakoshobby.cpp
#include <bits/stdc++.h>
using namespace std;

const int MOD = 1e9+7;
const int HMOD = 998244353;
const int MAXN = 1e5+5;
const double INF = 1e20;
const double EPS = 1e-9;

int getint() {
    int x;
    cin >> x;
    return x;
}

void getlist(vector<int> &lst, int n) {
    for (int i = 0; i < n; ++i) {
        cin >> lst[i];
    }
}

vector<int> getstr(int n) {
    vector<int> lst(n);
    for (int i = 0; i < n; ++i) {
        char c;
        cin >> c;
        lst[i] = c - '0';
    }
    return lst;
}

void findloop(vector<int>& memo, set<int>& vis, vector<int>& p, vector<int>& b, int idx) {
    if (memo[idx] != -1) return;
    if (vis.count(idx)) {
        int cur = p[idx], cnt = b[idx];
        while (cur != idx) {
            cnt += b[cur];
            cur = p[cur];
        }
        cur = p[idx];
        memo[idx] = cnt;
        while (cur != idx) {
            memo[cur] = cnt;
            cur = p[cur];
        }
        return;
    }
    vis.insert(idx);
    findloop(memo, vis, p, b, p[idx]);
}

int check(vector<int>& memo, vector<int>& p, vector<int>& b, int idx) {
    if (memo[idx] != -1) return memo[idx];
    memo[idx] = b[idx] + check(memo, p, b, p[idx]);
    return memo[idx];
}

void solve() {
    int n = getint();
    vector<int> p(n);
    getlist(p, n);
    for (int i = 0; i < n; ++i) {
        p[i]--;  // Adjust indexing from 1-based to 0-based
    }
    vector<int> b(n);
    string l; cin >> l;
    for (int i = 0; i < n; ++i) {
        b[i] = l[i] - '0';
        b[i] ^= 1;  // Perform XOR operation with 1
    }
    vector<int> memo(n, -1);
    for (int i = 0; i < n; ++i) {
        if (i == p[i]) {
            memo[i] = b[i];
            continue;
        }
        set<int> vis;
        findloop(memo, vis, p, b, i);
        check(memo, p, b, i);
    }
    for (int i = 0; i < n; ++i) {
        cout << memo[i] << " ";
    }
    cout << endl;
}

int main() {
    int testcases = 1;
    testcases = getint();
    for (int c = 1; c <= testcases; ++c) {
        solve();
    }
    return 0;
}

File: codeforces/sakurakoshobby.py
import random
import math
from collections import defaultdict, Counter, deque, OrderedDict
from queue import PriorityQueue
from heapq import heapify, heappush, heappop
from functools import lru_cache, reduce
from bisect import bisect_left, bisect_right
from types import GeneratorType
import sys

MOD = 10**9+7
HMOD = 998244353
MAXN = 10**5+5
INF = 1e20
EPS = 1e-9

sys.setrecursionlimit(int(2e5))

input = lambda : sys.stdin.readline().strip()
print = lambda *args : sys.stdout.write(" ".join(map(str, args)) + "\n")
write = lambda *args : sys.stdout.write(" ".join(map(str, args)))

getint = lambda : int(input())
getlist = lambda : list(map(int, input().split()))
getstr = lambda : list(input()) # mutable string

def findloop(memo, vis, p, b, idx):
    if memo[idx] != -1: return memo[idx]
    if idx in vis:
        cur, cnt = p[idx], b[idx]
        while cur != idx:
            cnt += b[cur]
            cur = p[cur]
        cur = p[idx]
        memo[idx] = cnt
        while cur != idx:
            memo[cur] = cnt
            cur = p[cur]
        return
    vis.add(idx)
    findloop(memo, vis, p, b, p[idx])

def check(memo, p, b, idx):
    if memo[idx] != -1: return memo[idx]
    memo[idx] = b[idx] + check(memo, p, b, p[idx])
    return memo[idx]

def solve():
    n = getint()
    p = [x - 1 for x in getlist()]
    b = [x ^ 1 for x in list(map(int, input()))]
    memo = [-1] * n
    for i in range(n):
        if i == p[i]:
            memo[i] = b[i]
            continue
        findloop(memo, set(), p, b, i)
        check(memo, p, b, i)
    print(*memo)
    

testcases = 1
testcases = getint()
for c in range(1, testcases+1):
    #write(f"Case {c}: ")
    solve()
